Pathing

rvagg opened the pull request:
Closes: https://github.com/filecoin-project/lassie/issues/82
This is on top of #116 for now.

This isn't complete because I don't have the integration tests I need to prove that it's doing the right thing with the different selectors--I'm particularly concerned about the sharded file and directory case for shallow fetch. Here's my WIP though and it appears to work on both cmdline and http daemon.

* `<CID>` or `<CID>/path` are supported in both cmdline and http request
* cmdline has an optional `--shallow` that will not do the full-depth DAG fetch at the end of the path - this is different to #82, but `?full` on the command line isn't really ideal (`?` a problem). I also think maybe full is the right default here, not shallow. But that's up for discussion!
* Optional http query parameter now of `"depthType"`, defaults to `"full"` but could also be `"shallow"`, has the same effect. Naming and defaults up for discussion of course, I just picked what I thought was sensible.

I ended up with a `types.NewRequestForPath` to make a new `RetrievalRequest` since it's being loaded up with properties and getting a bit unwieldly to create manually. This also lead to the current `Lassie#Fetch()` feeling a bit redundant, so I refactored away `Lassie#Retrieve()`, leaving just `Lassie#Fetch()` where you supply a `RetrievalRequest` that you pre-made.

Very basic client_test.go for a unit test I wanted to get in, but it has a lot of baggage since a lot of the infra is awkward to mock out; see lots of structs with empty methods in there. Perhaps this is useful to evolve over time beyond what I'm testing.

I wouldn't mind eyes on the shallow selector, I'll comment inline on that.

rvagg commented on pkg/retriever/selectorutils/selectorutils.go:
```diff
@@ -0,0 +1,76 @@
+			ssb.ExploreAll(ssb.ExploreRecursiveEdge()),
+		)
+	} else {
+		// ExploreAll only this node, interpreted as unixfs-preload, which will
+		// load sharded files and sharded directories, not go no further.
+		ss = ssb.ExploreInterpretAs("unixfs-preload", ssb.ExploreRecursive(
+			selector.RecursionLimitDepth(0),
+			ssb.ExploreAll(ssb.ExploreRecursiveEdge()),
+		))
```
Is there a better option here for this shallow selector? I could use just use a `ssb.ExploreInterpretAs("unixfs-preload", ssb.Matcher())` but I didn't really want to get matchers involved. But recursion depth 0 seems a bit overkill for just "explore the current node, interpreted as unixfs-preload".

/cc @willscott

hannahhoward replied:
I would use matcher, honestly.

hannahhoward commented on pkg/lassie/lassie.go:
```diff
@@ -91,36 +94,48 @@ func NewLassieWithConfig(ctx context.Context, cfg *LassieConfig) (*Lassie, error
 	}
 }
+
+// WithHost allows you to specify a custom libp2p host.
```
yay for all the new comments :) always appreciated

hannahhoward commented on pkg/retriever/graphsyncretriever.go:
```diff
@@ -405,12 +416,29 @@ func retrievalPhase(
 	eventsCallback datatransfer.Subscriber,
 ) (*types.RetrievalStats, error) {
+
+	if selector == nil {
```
it seems odd to have this code in each retriever. What if someone makes a new retriever and forgets? I'm kind of inclined to move it up somehow?

hannahhoward reviewed (commented):
Generally LGTM so far. A few nits listed, none super blocking.

rvagg commented:
I think this is ready now.

* Regarding the CI error captured above - it was a bug in the http server, it wasn't waiting for DT client setup to be complete. That's fixed now.
* Merged in the bitswap and graphsync http tests into a single parent test that can handle lots of variations
* Added lots of variations!

hannahhoward commented on go.mod:
```diff
@@ -5,53 +5,53 @@ go 1.19
 	github.com/ipfs/go-cid v0.3.2
 	github.com/ipfs/go-datastore v0.6.0
-	github.com/ipfs/go-graphsync v0.14.0
+	github.com/ipfs/go-graphsync v0.14.3-0.20230228013324-e6813b3847eb
```
can we cut a release of graphsync? It can just be 0.14.4

hannahhoward commented on pkg/client/client_test.go:
```diff
@@ -0,0 +1,306 @@
+	require.Nil(t, stats)
+}
+
+type fakeHost struct {
```
so much fake! it's ok, glad we actually unit tested a bit of this.

hannahhoward commented on pkg/internal/itest/http_fetch_test.go:
```diff
@@ -0,0 +1,574 @@
+package itest
```
this is.... awesome. 👍 💯 🥇 🙏 👏  + any other emojis the kids are into.

hannahhoward replied:
and I did actually read most of it. It's just great we can spin these up and test so succinctly

hannahhoward commented on pkg/internal/itest/http_fetch_test.go:
```diff
@@ -0,0 +1,574 @@
+package itest
+
+import (
```
not a now problem, but for the future, if we seperate the HTTP daemon, we'll want to make sure not to lose all these tests. I wonder given plans if we should consider at some point converting this to direct fetch. At the same time, HTTP API is the thing we're delivering for now, and I'd rather we focus our testing on that.

hannahhoward commented on go.mod:
```diff
@@ -5,53 +5,53 @@ go 1.19
 	github.com/ipfs/go-cid v0.3.2
 	github.com/ipfs/go-datastore v0.6.0
-	github.com/ipfs/go-graphsync v0.14.0
+	github.com/ipfs/go-graphsync v0.14.3
```
I went ahead adn pushed a tagged update to go-graphsync

hannahhoward reviewed (approved):
Weird to submit a "just LGTM" to such a long PR but this is really great.

hannahhoward commented on .github/workflows/go-check.yml:
```diff
@@ -8,26 +8,20 @@ jobs:
       - uses: actions/setup-go@v3
         with:
-          go-version: "1.19.x"
+          go-version: 1.20.x
```
aha! I thought 1.19.x was no longer the latest. I'll go ahead and put back in all my atomic.Uint's

rvagg commented:
* Added /path/to/*(exhaustive) tests - the ones in there were just shallow versions but now we test we can get full-DAG under a path
* Fixed a bug in the random unixfs directory generator that would let it generate directories significantly larger than what you requested
* Stopped the random unixfs directory generator making zero-sized directories or files
* Skipped the http daemon tests when the race detector is turned on, it's possible for them to run too slow since there's so many stacked up under a parent test and the race detector makes it so much slower on these CI machines
